lms/djangoapps/grades/tasks.py

```diff
@@ -41,67 +63,83 @@ def recalculate_subsection_grade(
+    course_key = CourseLocator.from_string(kwargs['course_id'])
     if not PersistentGradesEnabledFlag.feature_enabled(course_key):
         return
 
     score_deleted = kwargs['score_deleted']
+    scored_block_usage_key = UsageKey.from_string(kwargs['usage_id']).replace(course_key=course_key)
+    expected_modified_time = from_timestamp(kwargs['expected_modified_time'])
 
     # The request cache is not maintained on celery workers,
     # where this code runs. So we take the values from the
     # main request cache and store them in the local request
     # cache. This correlates model-level grading events with
     # higher-level ones.
-    set_event_transaction_id(kwargs.get('event_transaction_id', None))
-    set_event_transaction_type(kwargs.get('event_transaction_type', None))
-    scored_block_usage_key = UsageKey.from_string(usage_id).replace(course_key=course_key)
+    set_event_transaction_id(kwargs.pop('event_transaction_id', None))
+    set_event_transaction_type(kwargs.pop('event_transaction_type', None))
 
     # Verify the database has been updated with the scores when the task was
     # created. This race condition occurs if the transaction in the task
     # creator's process hasn't committed before the task initiates in the worker
     # process.
     if not _has_database_updated_with_new_score(
-            user_id, scored_block_usage_key, weighted_earned, weighted_possible, score_deleted,
+            kwargs['user_id'], scored_block_usage_key, expected_modified_time, score_deleted,
     ):
-        raise _retry_recalculate_subsection_grade(
-            user_id, course_id, usage_id, only_if_higher, weighted_earned, weighted_possible, score_deleted,
-        )
+        raise _retry_recalculate_subsection_grade(**kwargs)
 
     _update_subsection_grades(
         course_key,
         scored_block_usage_key,
-        only_if_higher,
-        course_id,
-        user_id,
-        usage_id,
-        weighted_earned,
-        weighted_possible,
+        kwargs['only_if_higher'],
+        kwargs['course_id'],
+        kwargs['user_id'],
+        kwargs['usage_id'],
+        kwargs['expected_modified_time'],
         score_deleted,
     )
 
 
 def _has_database_updated_with_new_score(
-        user_id, scored_block_usage_key, expected_raw_earned, expected_raw_possible, score_deleted,
+        user_id, scored_block_usage_key, expected_modified_time, score_deleted,
 ):
     """
     Returns whether the database has been updated with the
     expected new score values for the given problem and user.
-
-    Just here to let tests run while Eric updates his PR to go back
-    to timestamp-based comparison.
     """
-    return True
+    score = get_score(user_id, scored_block_usage_key)
+
+    if score is None:
+        # score should be None only if it was deleted.
+        # Otherwise, it hasn't yet been saved.
+        return score_deleted
+    elif score.module_type == 'openassessment':
```

Comment:
Changing this check from `score.grade is None` fixed my lettuce tests. LTI "helpfully" issues a score reset by sending a grade of `None, None`, which was triggering the check I had implemented before. Naturally, this failed quickly after as there was no submissions score to check. This check is much more accurate and will avoid the problem of non-openassessment problems hitting this code.